filebeat/input/filestream/input_test.go

```diff
@@ -0,0 +1,332 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package filestream
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	loginp "github.com/elastic/beats/v7/filebeat/input/filestream/internal/input-logfile"
+	input "github.com/elastic/beats/v7/filebeat/input/v2"
+	"github.com/elastic/beats/v7/libbeat/beat"
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/logp"
+	"github.com/elastic/beats/v7/libbeat/statestore"
+	"github.com/elastic/beats/v7/libbeat/statestore/backend/memlog"
+)
+
+// test_close_renamed from test_harvester.py
+func TestFilestreamCloseRenamed(t *testing.T) {
+	tmpDir, err := ioutil.TempDir("", "TestFilestreamCloseRenamed")
+	if err != nil {
+		t.Fatalf("cannot create temporary folder for test file: %+v", err)
+	}
+	defer os.RemoveAll(tmpDir)
+
+	testlogPath := filepath.Join(tmpDir, "test.log")
+	err = ioutil.WriteFile(testlogPath, []byte("first log line\n"), 0644)
+	if err != nil {
+		t.Fatalf("cannot write log lines to test file: %+v", err)
+	}
+
+	testStore := openTestStatestore(t, tmpDir)
+	inp := getInputFromConfig(
+		t,
+		testStore,
+		map[string]interface{}{
+			"paths":                                []string{testlogPath},
+			"prospector.scanner.check_interval":    "1s",
+			"close.on_state_change.check_interval": "1s",
+			"close.on_state_change.renamed":        "true",
+		},
+	)
+
+	ctx, cancelInput := context.WithCancel(context.Background())
+	inputCtx := input.Context{Logger: logp.L(), Cancelation: ctx}
+	pipeline := &mockPipelineConnector{}
+
+	var wg sync.WaitGroup
+	go func() {
+		wg.Add(1)
+		defer wg.Done()
+
+		inp.Run(inputCtx, pipeline)
+	}()
+
+	// first event has made it successfully
+	waitUntilEventCount(1, pipeline)
+	// check registry
+	checkOffsetInRegistry(t, testStore, testlogPath, len([]byte("first log line\n")))
+
+	testlogPathRenamed := filepath.Join(tmpDir, "test.log.rotated")
+	err = os.Rename(testlogPath, testlogPathRenamed)
+	if err != nil {
+		t.Fatalf("failed to rename file '%s': %+v", testlogPath, err)
+	}
```

Comment:
nit:

```
func mustRename(t *testing.T, from, to string) {
  ...
}
```

I would expect us to have quite a few scenarios that require a rename. Having a helper that also validates that the operation did succeed will make it easier to follow the flow.

As we operate on a test directory you can make it "independent" by providing a `TestDiskDriver` like:

```
type TestDiskDriver struct {
  t *testing.T
  rootPath string
}

func newTestDiskDriver(t *testing.T) *TestDiskDriver {
  return &TestDiskDriver{t: t, rootPath: t.Tempdir()}
}

func (d *TestDiskDriver) Open(path string) *os.File {
  f, err := os.Open(d.FullPath(path))
  require.NoError(t, err)
}

func (d *TestDiskDriver) WriteFile(path string, contents string) {
  err := ioutil.WriteFile(d.FullPath(path), []byte(contents))
  require.NoError(t, err)
}

func (d *TestDiskDriver) Rename(from, to string) {
  ...
}

func (d *TestDiskDriver) FullPath(path string) string {
  return filepath.Join(d.rootPath, path)
}
```